lib/collection/src/shards/transfer/stream_records.rs

```diff
@@ -24,18 +30,39 @@ pub(super) const TRANSFER_BATCH_SIZE: usize = 100;
 /// # Cancel safety
 ///
 /// This function is cancel safe.
+#[allow(clippy::too_many_arguments)]
 pub(super) async fn transfer_stream_records(
+    transfer_config: ShardTransfer,
     shard_holder: Arc<LockedShardHolder>,
     progress: Arc<Mutex<TransferTaskProgress>>,
     shard_id: ShardId,
     remote_shard: RemoteShard,
+    channel_service: &ChannelService,
+    consensus: &dyn ShardTransferConsensus,
     collection_id: &CollectionId,
-    filter: Option<Filter>,
 ) -> CollectionResult<()> {
     let remote_peer_id = remote_shard.peer_id;
     let cutoff;
+    let filter = transfer_config.filter;
     let merge_points = filter.is_some();
 
+    // Whether we need an intermediate replica state (ReadActive) during transfer to sync nodes
+    // We use this when transferring between different shard IDs to ensure data consistency, this
+    // way all readers can be switched to the new shard before any writers
+    let sync_intermediate_state = transfer_config
+        .to_shard_id
+        .is_some_and(|id| transfer_config.shard_id == id);
```

Comment:
```suggestion
        .is_some_and(|id| transfer_config.shard_id != id);
```

I think this is what you intended to have?